pkg/controller/job/job_controller.go

```diff
@@ -1549,13 +1572,150 @@ func getBackoff(queue workqueue.RateLimitingInterface, key interface{}) time.Dur
+		newBackoffInfo.lastFailureTime = nil
+	}
+
+	if j < len(newFailedPods) {
+		newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + int32(len(newFailedPods)-j)
+		lastFinshedTime := getFinishedTime(newFailedPods[len(newFailedPods)-1])
+		newBackoffInfo.lastFailureTime = &lastFinshedTime
+	}
+
+	return newBackoffInfo
+}
+
+func getInitialBackoffParameters(countedPods []*v1.Pod) (int32, *metav1.Time) {
+	sortByFinishedTime(countedPods)
+	failures := 0
+	var lastFailureTime *metav1.Time
+	for _, pod := range countedPods {
+		if pod.Status.Phase == v1.PodSucceeded {
+			break
+		}
+
+		failures = failures + 1
+		if lastFailureTime == nil {
+			finishedTime := getFinishedTime(pod)
+			lastFailureTime = &finishedTime
+		}
+	}
+
+	return int32(failures), lastFailureTime
+}
+
+func sortByFinishedTime(pods []*v1.Pod) {
+	finishedTimeSortFunction := func(p1, p2 *v1.Pod) bool {
+		p1FinishTime := getFinishedTime(p1)
+		p2FinishTime := getFinishedTime(p2)
+
+		return p1FinishTime.Before(&p2FinishTime)
+	}
+
+	sort.Slice(pods, func(i, j int) bool {
+		return finishedTimeSortFunction(pods[i], pods[j])
+	})
+}
+
+func getFinishedTime(p *v1.Pod) metav1.Time {
+	if p.ObjectMeta.DeletionTimestamp != nil {
+		return *p.ObjectMeta.DeletionTimestamp
+	}
+
+	finishTime := p.Status.ContainerStatuses[0].State.Terminated.FinishedAt
+	for _, cond := range p.Status.ContainerStatuses {
+		if finishTime.Before(&cond.State.Terminated.FinishedAt) {
+			finishTime = cond.State.Terminated.FinishedAt
+		}
+	}
+
+	return finishTime
+}
+
+func getCountedPods(pods, succeeded, failed []*v1.Pod) []*v1.Pod {
+	var countedPods []*v1.Pod
+
+	succeededSet := make(map[*v1.Pod]bool)
+	failedSet := make(map[*v1.Pod]bool)
+
+	for _, p := range succeeded {
+		succeededSet[p] = true
+	}
+
+	for _, p := range failed {
+		failedSet[p] = true
+	}
+
+	for _, p := range pods {
+		if _, ok := succeededSet[p]; ok {
+			continue
+		}
+
+		if _, ok := failedSet[p]; ok {
+			continue
+		}
+
+		if p.Status.Phase == v1.PodRunning || p.Status.Phase == v1.PodPending {
+			continue
+		}
+
+		countedPods = append(countedPods, p)
+	}
+
+	return countedPods
+}
+
+func getNewSucceededPods(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String) []*v1.Pod {
+	return getValidPodsWithFilter(job, pods, uncounted, expectedRmFinalizers, func(p *v1.Pod) bool {
+		return p.Status.Phase == v1.PodSucceeded
+	})
+}
+
+func getNewFailedPods(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String) []*v1.Pod {
+	return getValidPodsWithFilter(job, pods, uncounted, expectedRmFinalizers, func(p *v1.Pod) bool {
+		if feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) && job.Spec.PodFailurePolicy != nil {
+			if !isPodFailed(p, job) {
+				return false
+			}
+			_, countFailed, _ := matchPodFailurePolicy(job.Spec.PodFailurePolicy, p)
+			return countFailed
+		} else {
+			return isPodFailed(p, job)
+		}
+	})
+}
```

Comment:
This code could still be part of `getStatus` IMO.

Reply:
Done